proposals/3/README.md

```diff
@@ -0,0 +1,210 @@
+            *   EKS - IAM/role and IAM Policy with Route53 permissions., Kubernetes Service Account.
+        *   
+    *   _kaniko-sa_
+        *   required
+        *   IAM roles:	
+            *   GKE - roles/iam.workloadIdentityUser, roles/storage.admin, roles/storage.objectAdmin, roles/storage.objectCreator
+            *   EKS - Covered by the tekton-bot Service Account.
+    *   _storage-sa_
+        *   required
+        *   IAM roles:	
+            *   GKE - roles/iam.workloadIdentityUser, storage.objects.[create|get|delete|update]
+            *   EKS - S3Access Policy attached to an IAM/role and a Kubernetes Service Account.
+    *   _tekton-sa_
+        *   required
+        *   IAM roles:	
+            *   GKE - roles/iam.workloadIdentityUser, roles/viewer 
+            *   EKS - IAM/role and IAM Policy with general permissions, Kubernetes Service Account.
+    *   _velero-sa_
+        *   optional, only created if the user wants to use Velero
+        *   IAM roles:	
+            *   GKE - roles/iam.workloadIdentityUser, roles/storage.admin, roles/storage.objectAdmin, roles/storage.objectCreator
+            *   EKS - S3Access Policy attached to an IAM/role and a Kubernetes Service Account.
+    *   _vault-sa_
+        *   required            
+*   Storage buckets 
+    * Currently we have the following buckets
+        *   logs (required)
+        *   reports (optional, same bucket as logs if not provided)
+        *   repository (optional, used by bucketrepo if enabled)
+        *   backup (optional, used by Velero if enabled)
+    * To make things easier in the default Terraform setup we could consider using a single bucket.
+    This would simplify the conditional logic around bucket creation and would make it easier for the user to enable storage related features. 
+    This needs further investigation/discussion.
+*   Vault (optional, the user has the ability to BYO Vault instance)
+
+Permissions needed by Kubernetes service accounts for the various Cloud APIs are managed by [workload identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity).
+
+Similar to the setup in [terraform-google-jx](https://github.com/jenkins-x/terraform-google-jx), the output of applying the Terraform plan will be a `jx-requirements.yaml` which can then be used as input for `jx boot`.
+
+
+### 2.2.3 Boot and jx-requirements.yaml
+
+The plan is to leave the current Boot code, for now, untouched as much as possible.
+Boot needs to be provided with the URLs to the buckets to be used.
+These values need to be set in the existing `StorageEntryConfig.URL` fields. 
+
+For Vault, we will need to add a new field to `VaultConfig`, `VaultConfig.URL` pointing to the internal or external URL of the Vault instance to use.
+We can leave the other existing values or mark them as deprecated.
+
+`ClusterConfig` contains the already deprecated values `VaultName` and `VaultSAName`.
+It should be safe to remove them now. 
+
+
+### 2.2.4 Vault
```

Comment:
@rawlingsj, I seem to hear that you are not a big fan of Vault ;-) You seem to propose Google Secret Management which is fine, but how would you see this working across cloud providers. The benefit of Vault would be that you have some sort of generic secret management. WDYT?

Reply:
fwiw, feels to me like we should just be concerned with Vault _in the scope of this proposal_, since this is about separating out the cloud resource creation we do in boot currently. Adding support for other secret management systems would be a separate effort, one that I think would be notably easier once this effort is complete.

Reply:
>  I seem to hear that you are not a big fan of Vault ;-) 

correct! ;)

Main reason is, I'm not a fan of storing the secrets for the cluster, inside the cluster.  In a DR scenario having them outside is better if the cluster is lost.  With Google Secrets Manager, Workload Identity is used to provide a Kubernetes service account permission to super simply get secrets.  It's so nice and I'd expect AWS + Azure have similar approaches. 

In the spirit of "Using the Cloud Well" my vote would be to default to the managed cloud offerings where possible and fallback to Vault if not available or opted out.

Reply:
> In the spirit of "Using the Cloud Well" my vote would be to default to the managed cloud offerings where possible and fallback to Vault if not available or opted out.

well, if we still need Vault as the fallback option and given that Vault is what we have, I'd argue the natural approach is to start with Vault and take it from there. I guess we might be just disagreeing on the order things should be done. Would be nice if we could make the secret manager pluggable, but I think more heavy lifting is required for that first.

That said, I am intrigued by how you got Google Secrets Manager working? Are you saying you actually integrated this already in code? If so, do you have a pointer?

Reply:
> That said, I am intrigued by how you got Google Secrets Manager working? Are you saying you actually integrated this already in code? If so, do you have a pointer?

yes it's in the same installer we've been linking too ;) I was hoping folks might have tried it or looked at the docs as it is relevant to this enhancement https://github.com/jenkins-x-labs/jenkins-x-installer#secrets

Reply:
The point about secrets surviving outside the cluster seems really key to me. I can see the benefit there, and the benefit of removing the Vault complexity is also real.

What concerns me is that it's not particularly validated on the user side as a request yet, and the right way to do it seems to be making the secrets pluggable but that's quite a scope increase.

So, the call we'd have to make is between 

- Leaving things as-is for this initiative and reviewing secrets separately

- Building something bespoke for Google secrets and treating the rest of the cloud differently (which seems contrary to our tech debt goals here) 

- A significant scope increase to do this the right way

Reply:
I'd like to point out that secrets actually live outside the cluster right now using Vault.

Vault is actually configured to use buckets to store encrypted secrets with a KMS key.

If the cluster is destroyed for some reason, the secrets are going to be kept in the bucket until another Vault instance is created and is configured to use the same key and bucket.

The thing is that both AWS/GCP Secrets Manager are managed services and all the complexity is managed by either AWS or Google, which makes it easier to operate with and you'll probably get some kind of compromise from the providers part to keep your secrets up and secure, which is something that you have to do yourself if you use Vault.

Reply:
> I'd like to point out that secrets actually live outside the cluster right now using Vault.
Vault is actually configured to use buckets to store encrypted secrets with a KMS key.

Yes you're right, true.  Still favour the managed cloud provider solutions for the reasons you mention too.  Plus one less thing for Jenkins X to install + upgrade etc.

Reply:
I've also been there when restoring vault from a bucket, scary times ;) I seem to remember a Vault version change with the install once which caused a problem too.  Can't remember much more than that, maybe others can but all the same having secrets source of truth in a managed service is good.

Reply:
I think we should de-scope considering a different secret management system from this proposal. While one of the core aims of this proposal 'removing cloud resource creation responsibility from boot' would help move us in the direction of being able to use different secret storage systems I believe it is best to decouple it in to two different proposals. De-scoping adding additional secret management solutions should make it easier to successfully introduce this change to users and should also reduces the engineering effort required.

Reply:
> Plus one less thing for Jenkins X to install + upgrade etc.

Note, this proposal suggests doing the Vault creation (the actual instance) in Terraform. So `jx boot` gets as input via the requirements the URL to a Vault instance. The idea is that Boot does not have to create the Vault instance anymore.

Reply:
> I've also been there when restoring vault from a bucket, scary times ;) I seem to remember a Vault version change with the install once which caused a problem too. 

I know. I agree that it is a bit scary. Partly I think because we don't really have so much experience with running Vault in production and tackling problems like restoring etc.

Reply:
Vault is an interesting problem, is the proposition to manage this as a module, alongside JX?  With the state managed in the same place?  I'd be interested to see what terraform providers we are thinking of using (possibly should stay away from the helm provider, it doesn't handle updates properly).

Reply:
> Vault is an interesting problem,

Indeed

>  is the proposition to manage this as a module, alongside JX? 

Yes. Basically taking the bits and pieces from there https://github.com/sethvargo/vault-on-gke

> With the state managed in the same place?

Yes.

Reply:
> possibly should stay away from the helm provider, it doesn't handle updates properly

Good to know. We had been discussing the option of using the Helm Terraform provider to provision the Vault Operator, basically as an alternative to  https://github.com/sethvargo/vault-on-gke. But that might not be a good option then.